### ISSUE TITLE
Do you recommend running DEBIAS-M using GPU or CPU?

### ISSUE THREAD

**Issue by nettanetta** [korem-lab/DEBIAS-M#14]
Hi, 
Thank you for making DEBIAS-M publicly available and writing user-friendly documentation and API.
Do you recommend running it on GPU or CPU? Did you try both and can share an estimate of the runtime differences? 

Thank you
Netta

**Comment by gaustin15** [comment]
So far I’ve only run DEBIAS-M on CPUs and haven’t tested it on GPUs, and the option to use GPUs is currently not exposed in the main DEBIAS-M classes. In practice the runtime for these models is generally fast, even on CPU, so GPU acceleration hasn’t been necessary in the use cases I’ve worked with.

If there’s interest, I could add a push that exposes the torch option to select GPUs (it wouldn't take much work, probably 4 or 5 lines of code from my end). However, I don’t currently have easy access to GPUs to test this with, so it would be difficult for me to guarantee that the implementation works reliably without someone helping test it.

If GPU support would be useful for your workflow, please feel free to let me know; I’d be happy to explore adding it (I'm sure it would provide some level of speedup, by how much I can't say).

**Comment by nettanetta** [comment]
Thank you for your quick response, 

I'm trying to run DEBIAS-M (the online version) on the data from ["Gut Microbiome Wellness Index 2 enhances health status prediction from gut microbiome taxonomic profiles"](https://www.nature.com/articles/s41467-024-51651-9). The training set has ~8K samples from 54 studies (=batches). 
I would like to run some parameter search, but it currently took me over 24 hours to test 3 parameter configurations. 
Do you think GPU support can help enhance runtime? Is it possible to use multiple CPUs? Does this runtime surprise you? 
Thanks!
Netta

**Comment by gaustin15** [comment]
Thank you for the background information. That’s a pretty big benchmark - exciting! In general, can you perhaps share more information on what parameter sets you are testing DEBIAS-M with?

Re. ‘is this runtime surprising?’: it depends on what your testing structure is. If it involves doing something in a leave-one-study-out style that involves running DEBIAS-M ~54 separate times, with 3 parameter configs, a runtime >24hrs is in line with our benchmarks in our Fig. S10 and doesn’t seem all that crazy, we’re just working with a large dataset that involves many model iterations. 

If the analysis structure is something else, a helpful step (and perhaps useful in either evaluation structure anyways) would be to implement some level of minimum presence to remove features that are overwhelmingly zeros. This would help minimize some avoidable computation within debias (by allowing it to avoid carrying 1000s of zeros around). I (and others) would generally advocate for some level of minimum presence requirements anyways (eg. only keep taxa accounting for relative abundance > X in Y% of training set samples), both in analyses that do and don’t involve debias; although it’s of course difficult for me to guess about what processing variations are relevant to your particular analysis.

Re. your question about GPUs: I can expose it and I expect it to make things faster - we’ve never benchmarked it, so I’m not sure by how much. We would also need to work together on testing it. I think parallelizing over multiple CPUs as mentioned in (2) below is the more cost-efficient approach.

Re. using multiple CPUs: There’s two levels to my answer, it seems to me like the more relevant one to you is my second point.
1. If this is about scattering gradients within a single DEBIAS-M model across multiple CPUs during training, my answer will be similar to the GPU question. Torch does allow for this kind of thing to be implemented (e.g.torch.distributed DistributedDataParallel), although this wasn’t implemented within DEBIAS-M and, in my opinion, would not be worth implementing here. Primarily because…
2. If multiple CPUs are available, it is possible to scatter computation across different CPUs by running different parameter sets / different DEBIAS-M models (for each held out study) on the different CPUs in parallel. (e.g. if we are on a compute cluster, it is possible to submit separate jobs for each parameter set to separate compute nodes, and storing all results in some shared directory). Of course, this solution applies to models other than just DEBIAS-M, but I do believe that this is a simpler and more effective use of multiple CPUs than point (1). 



Hope this info helps.

**Comment by talkorem** [comment]
Hi, I'm closing this issue as there hasn't been any response for the past ~2 months - feel free to reach out again if we can help with anything else.
